File: python/meteroid/api/features.py
# this file is @generated
import typing as t

from ..models import (
    Feature,
    FeatureListResponse,
    FeatureStatus,
    ProductId,
)
from .common import ApiBaseAsync, ApiBaseSync, decode_response, serialize_query_params


class FeaturesAsync(ApiBaseAsync):
    """features API."""

    async def list_features(
        self,
        *,
        statuses: t.Optional[t.List[FeatureStatus]] = None,
        product_id: t.Optional[ProductId] = None,
        search: t.Optional[str] = None,
        page: t.Optional[int] = None,
        per_page: t.Optional[int] = None,
    ) -> FeatureListResponse:
        """:param statuses: Filter by feature status. Repeat the param to select multiple, omit to return all.
        :param product_id: Filter by product. Omit to return features across all products.
        :param search: Search by feature name.
        :param page: Page number (0-indexed)
        :param per_page: Number of items per page"""
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/features",
            query_params=serialize_query_params(
                {
                    "statuses": statuses,
                    "product_id": product_id,
                    "search": search,
                    "page": page,
                    "per_page": per_page,
                },
            ),
        )
        return decode_response(response, FeatureListResponse)

    async def get_feature(
        self,
        id_or_code: str,
    ) -> Feature:
        response = await self._request_asyncio(
            method="get",
            path="/api/v1/features/{id_or_code}",
            path_params={
                "id_or_code": id_or_code,
            },
        )
        return decode_response(response, Feature)


class Features(ApiBaseSync):
    """features API."""

    def list_features(
        self,
        *,
        statuses: t.Optional[t.List[FeatureStatus]] = None,
        product_id: t.Optional[ProductId] = None,
        search: t.Optional[str] = None,
        page: t.Optional[int] = None,
        per_page: t.Optional[int] = None,
    ) -> FeatureListResponse:
        """:param statuses: Filter by feature status. Repeat the param to select multiple, omit to return all.
        :param product_id: Filter by product. Omit to return features across all products.
        :param search: Search by feature name.
        :param page: Page number (0-indexed)
        :param per_page: Number of items per page"""
        response = self._request_sync(
            method="get",
            path="/api/v1/features",
            query_params=serialize_query_params(
                {
                    "statuses": statuses,
                    "product_id": product_id,
                    "search": search,
                    "page": page,
                    "per_page": per_page,
                },
            ),
        )
        return decode_response(response, FeatureListResponse)

    def get_feature(
        self,
        id_or_code: str,
    ) -> Feature:
        response = self._request_sync(
            method="get",
            path="/api/v1/features/{id_or_code}",
            path_params={
                "id_or_code": id_or_code,
            },
        )
        return decode_response(response, Feature)

File: go/features.go
// this file is @generated
package meteroid

import (
	"context"
	"net/http"
)

// FeaturesListFeaturesOptions carries the query and header parameters of
// Features.ListFeatures.
//
// Optional parameters are pointers; leave them nil to omit them.
type FeaturesListFeaturesOptions struct {
	// Filter by feature status. Repeat the param to select multiple, omit to return all.
	Statuses []FeatureStatus

	// Filter by product. Omit to return features across all products.
	ProductId *ProductId

	// Search by feature name.
	Search *string

	// Page number (0-indexed)
	Page *int32

	// Number of items per page
	PerPage *int32
}

// Features groups the features operations of the Meteroid API.
type Features struct {
	client *Client
}

func (a *Features) ListFeatures(ctx context.Context, options *FeaturesListFeaturesOptions) (*FeatureListResponse, error) {
	req := newRequest(http.MethodGet, "/api/v1/features")

	if options != nil {

		if len(options.Statuses) > 0 {
			for _, item := range options.Statuses {
				req.AddQueryParam("statuses", string(item))
			}
		}
		if options.ProductId != nil {
			req.SetQueryParam("product_id", string(*options.ProductId))
		}
		if options.Search != nil {
			req.SetQueryParam("search", *options.Search)
		}
		if options.Page != nil {
			req.SetQueryParam("page", formatInt(int64(*options.Page)))
		}
		if options.PerPage != nil {
			req.SetQueryParam("per_page", formatInt(int64(*options.PerPage)))
		}
	}

	var out FeatureListResponse
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

func (a *Features) GetFeature(ctx context.Context, idOrCode string) (*Feature, error) {
	req := newRequest(http.MethodGet, "/api/v1/features/{id_or_code}")
	req.SetPathParam("id_or_code", idOrCode)

	var out Feature
	if err := a.client.execute(ctx, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}
